New parameters for from() + unit tests + additions to vignette

torfason opened the pull request:
This pull request implements three new parameters for the `import::from()` function:

* `.all`: Import all objects from a library or module (if set to TRUE)
* `.except`: Omit one or more objects from the import (especially useful with `.all`)
* `.chdir`: Choose whether to change directory before sourcing modules
* `.directory`: Allow user to choose directory where R files for sourcing modules are located

It also adds unit tests to the package.

Finally, it expands the vignette to include documentation for `.character_only`

This version has been verified to build locally, to pass the unit tests, and to pass an R check.

mschilli87 commented on R/from.R:
```diff
@@ -59,7 +67,8 @@
 #' import::into("imports:parallel", makeCluster, parLapply, .from = parallel)
 from <- function(.from, ..., .into = "imports", .library = .libPaths()[1L],
-                 .character_only = FALSE)
+                 .all=FALSE, .except=character(),
```
l would suggest to change this to

```suggestion
                 .all=(length(.except) > 0), .except=character(),
```

so using `.except` implied `.all=TRUE` by default.
I don't see any use case for `.except` with `.all=FALSE` (and even if there is one, this change would still allow it by specifiying both parameters explicitely), so this will safe some code in most cases.

Also, personally I find

```R
import::from(foo, .except=c("bar", "bat"))
```

equally understandable as

```R
import::from(foo, .all=TRUE, .except=c("bar", "bat"))
```

but faster to parse.

torfason replied:
That’s a very nice touch! I hadn’t thought of that, I’ll update the pull request to include it.

smbache commented on R/from.R:
```diff
@@ -142,16 +151,35 @@ from <- function(.from, ..., .into = "imports", .library = .libPaths()[1L],
   } else {
     # Load the package namespace, which is passed to the import calls.
     spec <- package_specs(from)
+    all_objects <- getNamespaceExports(spec$pkg)
```
How would `.all`/`all_objects` work if `:::` notation is used?

torfason commented:
I had not thought of that but `getNamespaceExports()` only lists exported objects so my strong expectation would be that `import:::(mypackage, privatefun, .all=TRUE)` would import `privatefun`, plus all public functions, but not any other private functions. If the expectation would be that all private functions be imported as well we could use something like this trick to get a list of them:
https://stackoverflow.com/questions/8696158/find-all-functions-including-private-in-a-package

I could do that and add tests to make sure that the behavior is correct with `:::`.

smbache commented:
> I had not thought of that but `getNamespaceExports()` only lists exported objects so my strong expectation would be that `import:::(mypackage, privatefun, .all=TRUE)` would import `privatefun`, plus all public functions, but not any other private functions. If the expectation would be that all private functions be imported as well we could use something like this trick to get a list of them:
> https://stackoverflow.com/questions/8696158/find-all-functions-including-private-in-a-package
> 
> I could do that and add tests to make sure that the behavior is correct with `:::`.

An alternative is to make a decision, at least for now, that `.all/.except` is not compatible with `:::`, and raise if tried. This is a more conservative approach, as I think that using `:::` to get everything would not be advisable use in almost all cases?

torfason commented:
That sounds reasonable, I agree that importing private functions seems like a "you know what you are doing" feature and users could be expected to state all functions to import. They could even create a list programmatically themselves and pass it using `character_only`.

smbache commented:
> That sounds reasonable, I agree that importing private functions seems like a "you know what you are doing" feature and users could be expected to state all functions to import. They could even create a list programmatically themselves and pass it using `character_only`.

Sounds like a plan then :)

torfason commented:
I have added new parameter, `.directory`:

  * `.directory`: Allow user to choose directory where R files for sourcing modules are located

I have also fixed some issues, including problems with `import::here()` (pulling in fix on existing branch), and `.into=""`. Also, the environment is now always returned, improving python-like usage. This should fix quite a number of issues that are currently open.

There are also a number of additional unit tests, in a framework where new tests can be added by adding a new file in the `tests/test_import` directory and they get automatically run. The framework tries to replicate testthat, but some changes had to be done because the standard.

My suggestion would be to pull this request into the master branch of the official repo by @smbache, so that this becomes the official development version. I feel that this is very close to what an eventual 1.2.0 version could look like.

torfason commented:
I agree :-)

On Sun, Aug 16, 2020 at 3:42 AM Stefan Milton Bache <
notifications@github.com> wrote:

>
>
> *@smbache* approved this pull request.
>
>
>
>
> I say we should merge this in, it's already a big PR.
>
>
> Then decide on what's left we want to do before a submission and do that
> separately to this PR.
>
>
> Looks good - thanks for the effort!
>
>
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/smbache/import/pull/26#pullrequestreview-468031085>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYY3FSR4J3W2FV2AEGPNWTSA5IQPANCNFSM4ONMWFKA>
> .
>
>
> --

----------------------------------------------------------
  Magnus Thor Torfason
  Associate Professor, Innovation and Entrepreneurship
  University of Iceland
  Gimli G218 ( https://www.hi.is/gimli | https://z1.is/99PW42QX+JX )
  Tel: +354 - 525 4291
  Zoom: https://zoom.us/my/torfason
  Web: https://www.hi.is/~torfason
----------------------------------------------------------
